# Differential Gene Expression analysis

For this part, we will use `edgeR` program. You can read more about `edgeR` [here](http://bioconductor.org/packages/release/bioc/html/edgeR.html). `edgeR` supports differential expression analysis of RNA-seq expression profiles (count data) with biological replication. It implements a range of statistical methodology based on the negative binomial distributions, including empirical Bayes estimation, exact tests, generalized linear models and quasi-likelihood tests. This program can also be used for any other counts like genomic data generated from ChIP-seq, Bisulfite-seq, SAGE and CAGE.


## Install `edgeR`

Open the `RStudio` in your laptop and run these commands:

```
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("edgeR")
```

## Process counts data

Once complete, you are ready to run `edgeR` on your data. To simplify the analysis, we will process the count data such that we have each comparison that we want to run as a separate count file.

The counts data generated by the `featureCounts` has following headers:

```
Geneid  Chr     Start   End     Strand  Length  SRR6826996_starAligned.out.sam .... (11 more file names)
```

we will clean this up, so that we remove `_starAligned.out.sam` from the headers. We will also be not needing `Chr`, `Start`, `End`, `Strand`, and `Length`.

```
cut -f 1,7- count.txt |grep -v "^#" > counts_clean.txt
sed -i 's/_starAligned.out.sam//g' counts_clean.txt
```

The experiment layout was as follows:

| SRA_ID     | GEO_ID     | Condition    |
|------------|------------|--------------|
| SRR6826996 | GSM3038526 | Control 1    |
| SRR6826997 | GSM3038527 | Control 2    |
| SRR6826998 | GSM3038528 | Control 3    |
| SRR6826999 | GSM3038529 | 4hrs MPLA 1  |
| SRR6827000 | GSM3038530 | 4hrs MPLA 2  |
| SRR6827001 | GSM3038531 | 4hrs MPLA 3  |
| SRR6827002 | GSM3038532 | 24hrs MPLA 1 |
| SRR6827003 | GSM3038533 | 24hrs MPLA 2 |
| SRR6827004 | GSM3038534 | 24hrs MPLA 3 |
| SRR6827005 | GSM3038535 | 3dp MPLA 1   |
| SRR6827006 | GSM3038536 | 3dp MPLA 2   |
| SRR6827007 | GSM3038537 | 3dp MPLA 3   |


So, we will prepare individual files (counts), with
control vs. 4hrs, control vs. 24hrs and control vs. 3dp. If you are interested, you can also do other comparisons such as 4hrs vs. 24hrs, 4hrs vs. 3dp, 24hrs vs. 3dp.

First, take a look at the header:

```
head -n 1 counts_clean.txt |tr "\t" "\n" |nl
```

Now, use the combinations of `cut` or `awk` command to generate the desired comparison file.

eg:

```
awk '{print $1,$2,$3,$4,$5,$6,$7}' counts_clean.txt > control_4hrs.txt
```
or

```
cut -f 1-7 counts_clean.txt > control_4hrs.txt
```

generate all the files needed for next step (contro_24hrs.txt, control_3dp.txt)


## DGE


Below is the sample code for performing the analysis:


```
# import data
datain <- read.delim("control_4hrs.txt",row.names="Geneid")

# experimental design
DataGroups <- c("CTL", "CTL","CTL", "TRT", "TRT", "TRT")

# load edgeR
library(edgeR)

# create DGE object of edgeR
dgList <- DGEList(counts=datain,group=factor(DataGroups))

# filter data to retain genes that are represented at least 1 counts per million (cpm) in at least 2 samples
countsPerMillion <- cpm(dgList)
countCheck <- countsPerMillion > 1
keep <- which(rowSums(countCheck) >= 2)
dgList <- dgList[keep,]
dgList$samples$lib.size <- colSums(dgList$counts)

# normalization using TMM method
dgList <- calcNormFactors(dgList, method="TMM")

## data exploration
# MDS plot
png("plotmds.png")
plotMDS(dgList, method="bcv", col=as.numeric(dgList$samples$group))
dev.off()

# Dispersion estimates
design.mat <- model.matrix(~ 0 + dgList$samples$group)
colnames(design.mat) <- levels(dgList$samples$group)
dgList <- estimateGLMCommonDisp(dgList,design.mat)
dgList <- estimateGLMTrendedDisp(dgList,design.mat, method="power")
dgList <- estimateGLMTagwiseDisp(dgList,design.mat)
png("plotbcv.png")
plotBCV(dgList)
dev.off()

# Differentail expression analysis
fit <- glmFit(dgList, design.mat)
lrt <- glmLRT(fit, contrast=c(1,-1))
edgeR_results <- topTags(lrt, n=Inf)

# plot log2FC of genes and highlight the DE genes
deGenes <- decideTestsDGE(lrt, p=0.05)
deGenes <- rownames(lrt)[as.logical(deGenes)]
png("plotsmear.png")
plotSmear(lrt, de.tags=deGenes)
abline(h=c(-1, 1), col=2)
dev.off()
 
# save the results as a table
write.table(edgeR_results, file="Results_edgeR.txt")
```
